MQTT component

fabaff opened the pull request:
Ok, this is the current state of the MQTT component (just the working stuff is included). The received messages are printed to the command line and the outgoing ones are sent like the example below to a fix topic.

```
home-assistant/master <ServiceCall mqtt.mqtt_send: 11111=22222, service_call_id=140663655012672-1>
```

We have to decide how we want to handle the incoming messages. I think that it would be nice if messages sent to a topic like `home-assistant/switch/window1/` or `home-assistant/sensor/bedroom/` create automatically an entry in the frontend and display the current state/value. For distinguishing between sensors and switches I guess it would be sufficient to use the playload of the message for the value and an additional topic `../on` and `../off` for the switches.

Perhaps somebody can give some insight from other HA solutions.

There is a lot to do till this component is ready to include. Every help to bring this component forwards is appreciated.

fabaff committed:
add paho-mqtt

fabaff committed:
add mqtt component

fabaff committed:
add mqtt component

balloob commented on homeassistant/components/mqtt.py:
```diff
@@ -0,0 +1,164 @@
+            MQTT_CLIENT.run()
+
+        except ConnectionRefusedError:
+            _LOGGER.error("Can't connect to the broker. "
```
* Can you use `_LOGGER.exception` ? That way it will write the exception to the log too which might aid in debugging.

* Let's put a return after this error so that further code in this method is not executed.

fabaff replied:
Changed to `_LOGGER.exception`. I missed that or removed too much.

balloob commented on homeassistant/components/mqtt.py:
```diff
@@ -0,0 +1,164 @@
+
+    def received_mqtt(call):
+        """ Action when an MQTT message arrives. """
+        pass
```
I would suggest that any message that we received is send as an event over the bus. This will make it very easy for other components to subscribe.

_(please note that I am not familiar with the MQTT jargon at all and just made the attributes up)_

```python
EVENT_MQTT_MESSAGE_RECEIVED = 'MQTT_MESSAGE_RECEIVED'
hass.event.fire(EVENT_MQTT_MESSAGE_RECEIVED, {
  topic: […],
  subtopic: […],
  message: […],
})
```

fabaff replied:
There is no need for that. Removed.

fabaff replied:
How can I get around the problem that 'service_func' is a positional argument of register?
```
hass.services.register(DOMAIN, EVENT_MQTT_MESSAGE_RECEIVED)
```

balloob commented on homeassistant/components/mqtt.py:
```diff
@@ -0,0 +1,164 @@
+    def mqtt_on_message(self, mqttc, obj, msg):
+        """ Message callback """
+        self.msg = '{} {} {}'.format(msg.topic, str(msg.qos), str(msg.payload))
+        print(self.msg)
```
I would suggest that any message that we received is send as an event over the bus. This will make it very easy for other components to subscribe.

_(please note that I am not familiar with the MQTT jargon at all and just made the attributes up)_

```python
EVENT_MQTT_MESSAGE_RECEIVED = 'MQTT_MESSAGE_RECEIVED'
hass.event.fire(EVENT_MQTT_MESSAGE_RECEIVED, {
  topic: […],
  subtopic: […],
  message: […],
})
```

balloob commented on homeassistant/components/mqtt.py:
```diff
@@ -0,0 +1,164 @@
+
+    def mqtt_on_message(self, mqttc, obj, msg):
+        """ Message callback """
+        self.msg = '{} {} {}'.format(msg.topic, str(msg.qos), str(msg.payload))
```
this line can be removed after we send it over the bus.

balloob commented:
Let me just say that for now, the current implementation seems like a good MVP (minimum viable product). Some minor comments but it should be good to go.

For the future, I think that these are different use cases for MQTT:
 1. Allow components in HA to subscribe to any channel on the MQTT server
 2. Allow Home Assistant to be controlled over MQTT (specific implementation of 1)
 3. Allow components in HA to control devices over MQTT by sending messages to any topic

balloob commented:
@fabaff do you have time to wrap up this PR? I would like to be able to put out a release this week.

balloob committed:
Sun component will work now without internet

balloob committed:
Better error reporting remote classes

balloob committed:
Reorg tests folder

balloob committed:
Extract helpers.event from core + misc cleanup

fabaff committed:
commit for merge

fabaff committed:
for merge

balloob committed:
Minor cleanup core

balloob committed:
Built-in components no longer use deprecated methods

balloob committed:
Built-in component cleanup

balloob committed:
Clean up test code

balloob committed:
Remove more deprecated method calls

balloob committed:
Minor code cleanup

balloob committed:
Nicer test imports

unittest changes import path so old style worked but is confusing

balloob committed:
Extended test_init tests to cover all

balloob committed:
Remove deprecated code

balloob committed:
Remove support for old home-assistant.conf file

balloob committed:
Update coveragerc

balloob committed:
Merge branch 'dev' into cleanup

Conflicts:
	.coveragerc

persandstrom committed:
Exception when client not in leases

Client can be dropped from leases list from where the status is
retreived before it is dropped from the ip neigh list.
The client needs a default status.

balloob committed:
Merge pull request #249 from persandstrom/asuswrt

ASUSWRT - Exception when client not in leases

balloob committed:
Merge pull request #239 from gbarba/rpi_gpio_component

Add rpi_gpio platform to switch

balloob committed:
Merge pull request #244 from balloob/cleanup

Core Cleanup

fabaff committed:
Merge branch 'mqtt' of github.com:fabaff/home-assistant into mqtt

fabaff committed:
add suggestions for discussion

fabaff commented:
So I commit the not-working version because I think that it will be easier to spot my mistakes.

balloob commented:
It should be `hass.bus.fire`, my bad.

fabaff commented:
Preview of how the MQTT messages are looking:
```
15-08-07 17:01:48 INFO (Thread-17) [homeassistant] Bus:Handling <Event MQTT_MESSAGE_RECEIVED[L]: topic=home-assistant/switch/1/on, subtopic=test, payload=b'Switch is ON', qos=0>
```

fabaff committed:
fix pylint warning and change to hass.bus.fire

balloob commented on homeassistant/components/mqtt.py:
```diff
@@ -0,0 +1,170 @@
+            'topic': msg.topic,
+            'subtopic': 'test',
+            'qos': str(msg.qos),
+            'payload': str(msg.payload),
```
Instead of `str(msg.payload)` use `msg.payload.decode('utf-8')`, same for `qos`

fabaff replied:
Thanks

balloob commented on homeassistant/components/mqtt.py:
```diff
@@ -0,0 +1,170 @@
+        self.msg = '{} {} {}'.format(msg.topic, str(msg.qos), str(msg.payload))
+        self.hass.bus.fire(EVENT_MQTT_MESSAGE_RECEIVED, {
+            'topic': msg.topic,
+            'subtopic': 'test',
```
Should this subtopic be hardcoded to `test` ?

fabaff replied:
No, was just a test...will be removed. First I was thinking about handling the topic here but it seems to make more sense to do that later in the various platforms.

balloob commented on homeassistant/components/mqtt.py:
```diff
@@ -0,0 +1,170 @@
+
+    def mqtt_on_message(self, mqttc, obj, msg):
+        """ Message callback """
+        self.msg = '{} {} {}'.format(msg.topic, str(msg.qos), str(msg.payload))
```
It seems that `self.msg` is never used and can be removed.

fabaff replied:
Yes, was just there to print the message to the command line for debugging.

fabaff committed:
update header and import

fabaff committed:
remove subtopic and convert payload to str

fabaff committed:
remove import

fabaff committed:
convert qos to str and remove msg
